satpy/readers/utils.py

```diff
@@ -210,57 +230,84 @@ def unzip_file(filename, prefix=None):
+        return None
+
+    fdn, tmpfilepath = tempfile.mkstemp(prefix=prefix,
+                                        dir=config["tmp_dir"])
+    LOGGER.info("Using temp file for BZ2 decompression: %s", tmpfilepath)
+    # try pbzip2
+    pbzip = which('pbzip2')
+    # Run external pbzip2
+    if pbzip is not None:
+        n_thr = os.environ.get('OMP_NUM_THREADS')
+        if n_thr:
+            runner = [pbzip,
+                      '-dc',
+                      '-p'+str(n_thr),
+                      filename]
+        else:
+            runner = [pbzip,
+                      '-dc',
+                      filename]
+        p = Popen(runner, stdout=PIPE, stderr=PIPE)  # nosec
+        stdout = BytesIO(p.communicate()[0])
+        status = p.returncode
+        if status != 0:
+            raise IOError("pbzip2 error '%s', failed, status=%d"
+                          % (filename, status))
         with closing(os.fdopen(fdn, 'wb')) as ofpt:
             try:
-                ofpt.write(bz2file.read())
+                stdout.seek(0)
+                shutil.copyfileobj(stdout, ofpt)
             except IOError:
-                import traceback
-                traceback.print_exc()
-                LOGGER.info("Failed to read bzipped file %s", str(filename))
+                LOGGER.debug("Failed to read bzipped file %s", str(filename))
                 os.remove(tmpfilepath)
-                return None
+                raise
         return tmpfilepath
+    # Otherwise, fall back to the original method
+    bz2file = bz2.BZ2File(filename)
+    with closing(os.fdopen(fdn, 'wb')) as ofpt:
+        try:
+            ofpt.write(bz2file.read())
+        except IOError:
+            LOGGER.debug("Failed to read bzipped file %s", str(filename))
+            os.remove(tmpfilepath)
+            return None
+    return tmpfilepath
 
-    return None
+
+def _unzip_FSFile(filename: FSFile, prefix=None):
+    """Open and Unzip remote FSFile ending with 'bz2'.
+
+    Args:
+        filename: The FSFile to unzip.
+        prefix (str, optional): If file is one of many segments of data, prefix random filename
+        for correct sorting. This is normally the segment number.
+
+    Returns:
+        Temporary filename path for decompressed file or None.
+
+    """
+    if not str(filename).endswith('bz2'):
+        return None
+
+    fdn, tmpfilepath = tempfile.mkstemp(prefix=prefix,
+                                        dir=config["tmp_dir"])
+    # open file
+    content = filename.open().read()
+    # unzip file if zipped (header start with hex 425A68)
+    if content.startswith(bytes.fromhex("425A68")):
+        content = bz2.decompress(content)
+
+    with closing(os.fdopen(fdn, 'wb')) as ofpt:
+        try:
+            ofpt.write(content)
+        except IOError:
+            LOGGER.debug("Failed to read bzipped file %s", str(filename))
+            os.remove(tmpfilepath)
+            return None
+    return tmpfilepath
```

Comment:
This block of code is shared between the two helper functions. Could it be extracted and the caller just provides the `content` bytes? Oh I guess it would also need `fdn`, `filename`, and `tmpfilepath`. That's unfortunate, but I think the cleaner and less duplicated code is worth it. I think it could be called like:

```python
return _write_uncompressed_bz2_data(content, fdn, filename, tmpfilepath)
```

Reply:
Side note: Would the memory saving be worth it to decompress blocks of data at a time? Basically open the destination file, read N bytes of data, decompress them, write the decompressed data, then loop again and read N bytes of compressed data, decompress, and so on.

@simonrp84 thoughts? Maybe in another PR?